src/server/scripts/Zandalar/AtalDazar/boss_volkaal.cpp

```diff
@@ -0,0 +1,417 @@
+            me->GetCreatureListWithEntryInGrid(totems, me->GetEntry());
+            for (Creature* totem : totems)
+            {
+                totem->SetUnkillable(false);
+                totem->KillSelf();
+            }
+        }
+    }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        InstanceScript* instance = me->GetInstanceScript();
+        if (!instance)
+            return;
+
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+    }
+};
+
+// 250372 - Lingering Nausea
+class spell_volkaal_lingering_nausea : public AuraScript
+{
+    void HandlePeriodic(AuraEffect const* aurEff) const
+    {
+        Unit* target = GetTarget();
+        if (!target)
+            return;
+
+        GetCaster()->CastSpell(target, aurEff->GetAmount(), CastSpellExtraArgsInit{
+            .TriggerFlags = TRIGGERED_IGNORE_CAST_IN_PROGRESS | TRIGGERED_DONT_REPORT_CAST_ERROR,
+            .TriggeringAura = aurEff
+        });
+    }
+
+    void Register() override
+    {
+        OnEffectPeriodic += AuraEffectPeriodicFn(spell_volkaal_lingering_nausea::HandlePeriodic, EFFECT_0, SPELL_AURA_PERIODIC_DUMMY);
+    }
+};
+
+// 250368 - Noxious Stench
+class spell_volkaal_noxious_stench : public SpellScript
+{
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo ({ SPELL_LINGERING_NAUSEA });
+    }
+
+    void HandleDummy(SpellEffIndex /*effIndex*/) const
+    {
+        GetCaster()->CastSpell(GetHitUnit(), SPELL_LINGERING_NAUSEA, CastSpellExtraArgsInit{
+            .TriggerFlags = TRIGGERED_IGNORE_CAST_IN_PROGRESS | TRIGGERED_DONT_REPORT_CAST_ERROR,
+            .TriggeringSpell = GetSpell()
+        });
+    }
+
+    void Register() override
+    {
+        OnEffectHitTarget += SpellEffectFn(spell_volkaal_noxious_stench::HandleDummy, EFFECT_0, SPELL_EFFECT_SCHOOL_DAMAGE);
+    }
+};
+
+// 250708 - Toxic Leap
+class spell_volkaal_toxic_leap_selector : public SpellScript
+{
+    void HandleScript(SpellEffIndex /*effIndex*/) const
+    {
+        GetCaster()->CastSpell(GetHitUnit(), GetEffectValue(), CastSpellExtraArgsInit{
+            .TriggerFlags = TRIGGERED_IGNORE_CAST_IN_PROGRESS | TRIGGERED_DONT_REPORT_CAST_ERROR,
+            .TriggeringSpell = GetSpell(),
+        });
+    }
+
+    void Register() override
+    {
+        OnEffectHitTarget += SpellEffectFn(spell_volkaal_toxic_leap_selector::HandleScript, EFFECT_0, SPELL_EFFECT_DUMMY);
+    }
+};
+
+// 250229 - Soul Anchor
+class spell_volkaal_soul_anchor : public SpellScript
+{
+    void SelectTotem(std::list<WorldObject*>& targets) const
+    {
+        if (targets.empty())
+        {
+            InstanceScript* instance = GetCaster()->GetInstanceScript();
+            if (!instance)
+                return;
+
+            if (Creature* totem = instance->GetCreature(DATA_VOLKAAL_TOTEM))
+                if (npc_volkaal_reanimation_totem* totemAI = CAST_AI(npc_volkaal_reanimation_totem, totem->GetAI()))
```

Comment:
no need to cast to script ai itself, for DoAction you can just do `totem->AI()->DoAction`

(execute action on boss when DoAction part changed as stated above)